mlir/lib/Dialect/MIOpen/Transforms/LowerMIOpenOps.cpp

```diff
@@ -107,13 +107,896 @@ const ArgumentFields Conv2DRewritePattern<miopen::Conv2DBwdWeightOp>::fields = {
+                           {gemmKBlocks, n / gemmKBlocks});
+    firstTransform.passThrough({"ko", "ho", "wo"}, {3, 4, 5},
+                               {"ko", "ho", "wo"});
+
+    TransformMapAttr firstTransformAttr = firstTransform.get();
+    Value transformed =
+        b.create<miopen::TransformOp>(loc, op.output(), firstTransformAttr);
+
+    // Map G and N0 to gemmG, N1HW to gemmK and K to gemmM
+    auto gemmTransform =
+        BottomUpCTBuilder::above(firstTransform, firstTransformAttr);
+    gemmTransform.merge("gemmG", 0, {"go", "n0"});
+    gemmTransform.merge("gemmK", 1, {"n1", "ho", "wo"});
+    gemmTransform.passThrough({"gemmM"}, {2}, {"ko"});
+
+    TransformMapAttr gemmTransformAttr = gemmTransform.get();
+    gemmOutput =
+        b.create<miopen::TransformOp>(loc, transformed, gemmTransformAttr);
+  }
+
+  // Set attributes for gridwise_gemm op.
+  llvm::SmallVector<NamedAttribute, 8> gridwiseGemmAttrs{
+      b.getNamedAttr("gemm_id", gemmIdAttr), b.getNamedAttr("arch", archAttr),
+      b.getNamedAttr("num_cu", numCuAttr)};
+
+  // TODO(kdrewnia) We can't use vector stores since this operation
+  // transforms KYXC outputs to KCYX ones. Until we stop doing the twist,
+  // we can't vectorize the writes
+  if (auto gemmVectorization =
+          op->template getAttrOfType<IntegerAttr>("matrix_c_data_per_copy")) {
+    op->setAttr("matrix_c_data_per_copy", b.getI32IntegerAttr(1));
+  }
+
+  // xdlopsV2.
+  auto xdlopsV2Attr = op->template getAttrOfType<BoolAttr>("xdlopsV2");
+  bool isXdlops = (xdlopsV2Attr && xdlopsV2Attr.getValue() == true);
+  if (isXdlops)
+    gridwiseGemmAttrs.push_back(
+        b.getNamedAttr("xdlopsV2", b.getBoolAttr(true)));
+
+  gridwiseGemmAttrs.push_back(b.getNamedAttr(
+      "kernel_algorithm", b.getStringAttr("backward_weight_v4r4")));
+
+  // This kernel is not run when there is padding on the GEMM
+  auto paddingInfo =
+      PaddingInfoAttr::get(b.getContext(), 0, 0, 0, BwdPaddingKernelInfo::NA);
+  auto dataOperation = InMemoryDataOperation::AtomicAdd;
+
+  Value gemmA = gemmOutput;
+  ArrayAttr oobA =
+      getBoundsCheckAttr(b, outputOobCheckDims, outputShape.size());
+  Value gemmB = gemmInput;
+  ArrayAttr oobB = getBoundsCheckAttr(b, inputOobCheckDims, inputShape.size());
+  Value gemmC = gemmFilter;
+  ArrayAttr oobC =
+      getBoundsCheckAttr(b, filterOobCheckDims, filterShape.size());
+
+  if (isXdlops) {
+    auto gop = b.create<miopen::GridwiseGemmV2Op>(
+        loc, gemmA, gemmB, gemmC, oobA, oobB, oobC, paddingInfo, dataOperation,
+        gridwiseGemmAttrs);
+    affixGridwiseGemmAttributes(op, gop, b);
+  } else {
+    op->emitOpError("Backward weight atomic add kernel requires xdlops and "
+                    "shouldn't have been tried without them");
+  }
+
+  // Finally, erase the original Conv2D op.
+  op.erase();
+
+  return success();
+}
+
+LogicalResult backwardData(miopen::Conv2DBwdDataOp op, PatternRewriter &b) {
```

Comment:
Can we safely assume the logic of the removed file for backward data padding kernel is located here?

Reply:
Yep